File: analytics/src/test/java/io/appmetrica/analytics/impl/CrashReporterTest.java
package io.appmetrica.analytics.impl;

import io.appmetrica.analytics.AppMetricaConfig;
import io.appmetrica.analytics.impl.client.ProcessConfiguration;
import io.appmetrica.analytics.impl.crash.CrashToFileWriter;
import io.appmetrica.analytics.impl.crash.UnhandledExceptionEventFormer;
import io.appmetrica.analytics.impl.crash.client.UnhandledException;
import io.appmetrica.analytics.impl.reporter.CrashReporterContext;
import io.appmetrica.analytics.impl.reporter.ReporterLifecycleListener;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.robolectric.RobolectricTestRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class CrashReporterTest extends BaseReporterTest {

    @Mock
    private CrashToFileWriter mCrashToFileWriter;
    @Mock
    private UnhandledExceptionEventFormer mEventFormer;
    @Mock
    private ReporterLifecycleListener reporterLifecycleListener;
    private CrashReporter mCrashReporter;

    @Override
    protected BaseReporter getReporter() {
        return new CrashReporter(
                mContext,
                mReportsHandler,
                mReporterEnvironment,
                mExtraMetaInfoRetriever,
                mCrashToFileWriter,
                mEventFormer,
                processDetector,
                unhandledExceptionConverter,
                regularErrorConverter,
                customErrorConverter,
                anrConverter,
                pluginErrorDetailsConverter
        );
    }

    @Before
    public void setUp() {
        super.setUp();
        mCrashReporter = (CrashReporter) mReporter;
    }

    @Test
    public void testMainReporterListenerIsCalledInConstructor() {
        when(ClientServiceLocator.getInstance().getReporterLifecycleListener())
                .thenReturn(reporterLifecycleListener);
        final BaseReporter reporter = getReporter();
        verify(reporterLifecycleListener).onCreateCrashReporter(any(CrashReporterContext.class));
    }

    @Test
    public void reportUnhandledException() {
        UnhandledException unhandledException = mock(UnhandledException.class);
        ReportToSend reportToSend = mock(ReportToSend.class);
        when(mEventFormer.formEvent(unhandledException, mReporterEnvironment)).thenReturn(reportToSend);
        mReporter.reportUnhandledException(unhandledException);
        verify(mCrashToFileWriter).writeToFile(reportToSend);
    }

    @Test
    public void reporterType() {
        AppMetricaConfig config = AppMetricaConfig.newConfigBuilder(apiKey).build();
        mReporter = new CrashReporter(
                mContext,
                mock(ProcessConfiguration.class),
                config,
                mReportsHandler
        );
        assertThat(mReporter.getEnvironment().getReporterConfiguration().getReporterType()).isEqualTo(CounterConfigurationReporterType.CRASH);
    }

    @Test
    public void updateConfigNullErrorEnvironment() {
        mCrashReporter.updateConfig(AppMetricaConfig.newConfigBuilder(apiKey).build());
        verify(mReporterEnvironment, never()).putErrorEnvironmentValue(anyString(), anyString());
    }

    @Test
    public void updateConfigFilledErrorEnvironment() {
        String key1 = "key1";
        String value1 = "value1";
        String key2 = "key2";
        String value2 = "value2";
        mCrashReporter.updateConfig(
                AppMetricaConfig.newConfigBuilder(apiKey)
                        .withErrorEnvironmentValue(key1, value1)
                        .withErrorEnvironmentValue(key2, value2)
                        .build()
        );
        verify(mReporterEnvironment).putErrorEnvironmentValue(key1, value1);
        verify(mReporterEnvironment).putErrorEnvironmentValue(key2, value2);
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/CrashReporterTest.kt
package io.appmetrica.analytics.impl

import io.appmetrica.analytics.AppMetricaConfig
import io.appmetrica.analytics.coreutils.internal.logger.LoggerStorage
import io.appmetrica.analytics.impl.crash.jvm.client.ThrowableModel
import io.appmetrica.analytics.impl.crash.jvm.client.UnhandledException
import io.appmetrica.analytics.impl.reporter.CrashReporterContext
import io.appmetrica.analytics.logger.appmetrica.internal.PublicLogger
import io.appmetrica.analytics.testutils.ClientServiceLocatorRule
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.on
import io.appmetrica.analytics.testutils.staticRule
import org.junit.Rule
import org.junit.Test
import org.mockito.kotlin.clearInvocations
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoInteractions
import org.mockito.kotlin.whenever
import java.util.UUID

internal class CrashReporterTest : CommonTest() {

    private val apiKey = UUID.randomUUID().toString()
    private val oneMoreApiKey = UUID.randomUUID().toString()
    private val reporterEnvironment: ReporterEnvironment = mock()
    private val cashReporterContext: CrashReporterContext = mock()
    private val reportsHandler: ReportsHandler = mock()

    private val crashReporterFieldsProvider: CrashReporterFieldsProvider = mock {
        on { reporterEnvironment } doReturn reporterEnvironment
        on { crashReporterContext } doReturn cashReporterContext
        on { reportsHandler } doReturn reportsHandler
    }

    private val throwable: ThrowableModel = mock()
    private val unhandledException = UnhandledException(
        throwable,
        null,
        null,
        null,
        null,
        null,
        null,
        null
    )

    private val firstEnvironmentKey = "Environment key #1"
    private val firstEnvironmentValue = "Environment value #1"
    private val secondEnvironmentKey = "Environment key #2"
    private val secondEnvironmentValue = "Environment value #2"

    private val config = AppMetricaConfig.newConfigBuilder(UUID.randomUUID().toString())
        .withErrorEnvironmentValue(firstEnvironmentKey, firstEnvironmentValue)
        .withErrorEnvironmentValue(secondEnvironmentKey, secondEnvironmentValue)
        .build()

    @get:Rule
    val clientServiceLocatorRule = ClientServiceLocatorRule()

    private val logger: PublicLogger = mock()

    @get:Rule
    val loggerStorageMockedStaticRule = staticRule<LoggerStorage> {
        on { LoggerStorage.getOrCreatePublicLogger(apiKey) } doReturn logger
        on { LoggerStorage.getOrCreatePublicLogger(oneMoreApiKey) } doReturn logger
        on { LoggerStorage.getMainPublicOrAnonymousLogger() } doReturn logger
    }

    private val crashReporter by setUp { CrashReporter(crashReporterFieldsProvider) }

    @Test
    fun constructor() {
        verify(ClientServiceLocator.getInstance().reporterLifecycleListener)!!
            .onCreateCrashReporter(cashReporterContext)
    }

    @Test
    fun `constructor if no reporter lifecycle listener`() {
        clearInvocations(ClientServiceLocator.getInstance().reporterLifecycleListener)
        whenever(ClientServiceLocator.getInstance().reporterLifecycleListener).thenReturn(null)
        CrashReporter(crashReporterFieldsProvider)
    }

    @Test
    fun reportUnhandledException() {
        crashReporter.reportUnhandledException(unhandledException)
        verify(reportsHandler).reportCrash(unhandledException, reporterEnvironment)
        verify(logger).info("Unhandled exception received: $unhandledException")
    }

    @Test
    fun `updateConfig with errorEnvironment`() {
        crashReporter.updateConfig(config)
        verify(reporterEnvironment).putErrorEnvironmentValue(firstEnvironmentKey, firstEnvironmentValue)
        verify(reporterEnvironment).putErrorEnvironmentValue(secondEnvironmentKey, secondEnvironmentValue)
    }

    @Test
    fun `updateConfig without errorEnvironment`() {
        AppMetricaConfig.newConfigBuilder(apiKey).build()
        crashReporter.updateConfig(AppMetricaConfig.newConfigBuilder(oneMoreApiKey).build())
        verifyNoInteractions(reporterEnvironment)
    }
}
